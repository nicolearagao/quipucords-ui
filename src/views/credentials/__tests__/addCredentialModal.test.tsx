import React, { act } from 'react';
import { render, renderHook, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { shallowComponent } from '../../../../config/jest.setupTests';
import helpers from '../../../helpers';
import { AddCredentialModal, CredentialFormFields, useCredentialForm } from '../addCredentialModal';

describe('AddCredentialModal', () => {
  let mockOnClose;
  let mockOnSubmit;

  beforeEach(() => {
    mockOnClose = jest.fn();
    mockOnSubmit = jest.fn();
    render(
      <AddCredentialModal
        isOpen={true}
        credentialType="network"
        credential={undefined}
        onClose={mockOnClose}
        onSubmit={mockOnSubmit}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const component = await shallowComponent(<AddCredentialModal isOpen={true} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should have the correct title', () => {
    const title = screen.getByText('Add Credential: network');
    expect(title).toMatchSnapshot('AddCredentialModal Title');
  });

  it('should call onSubmit with the correct filtered data when "Save" is clicked', async () => {
    const user = userEvent.setup();
    await user.type(screen.getByPlaceholderText('Enter a name for the credential'), 'Test Credential');
    await user.click(screen.getByText('Save'));

    expect(mockOnSubmit.mock.calls).toMatchSnapshot('onSubmit, filtered data');
  });

  it('should call onClose', async () => {
    const user = userEvent.setup();
    await user.click(screen.getByText('Cancel'));

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});

describe('useCredentialForm', () => {
  it('should initialize formData correctly', () => {
    const { current } = renderHook(() => useCredentialForm('network')).result;
    expect(current.formData).toMatchSnapshot('formData');
  });

  it('should derive token auth for specific credential types', () => {
    const { current: openshift } = renderHook(() => useCredentialForm('openshift')).result;
    const { current: rhacs } = renderHook(() => useCredentialForm('rhacs')).result;

    expect({
      openshift: openshift.authType,
      rhacs: rhacs.authType
    }).toMatchSnapshot('token auth');
  });

  it('should derive auth type based on the provided credential', () => {
    const mockCredential = {
      id: 1,
      name: 'Test Credential',
      username: 'admin',
      password: '',
      cred_type: 'network',
      created_at: new Date(),
      updated_at: new Date(),
      ssh_key: 'ssh-key',
      auth_token: '',
      ssh_passphrase: '',
      become_method: '',
      become_password: '',
      become_user: '',
      sources: []
    };
    jest.spyOn(helpers, 'getAuthType').mockReturnValue(helpers.authType.SSHKey);
    const { current: credentialBased } = renderHook(() => useCredentialForm('network', mockCredential)).result;
    expect(credentialBased.authType).toBe('SSH Key');
    expect(credentialBased.authType).toMatchSnapshot('auth type based on credential');
  });

  it('should update and filter formData when handleInputChange is called', () => {
    const result = renderHook(() => useCredentialForm('network')).result;
    const mockValue = 'Lorem ipsum';

    act(() => result.current.handleInputChange('name', mockValue));
    expect(result.current.formData.name).toBe(mockValue);

    const filteredData = result.current.filterFormData(result.current.formData);
    expect(filteredData.name).toBe(mockValue);
  });
});

describe('CredentialFormFields', () => {
  it('should render a basic component', async () => {
    const component = await shallowComponent(<CredentialFormFields />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should render specific to authType for type network', async () => {
    const networkUsernamePassword = await shallowComponent(
      <CredentialFormFields typeValue="network" authType="Username and Password" />
    );
    expect(networkUsernamePassword).toMatchSnapshot('network, Username and Password');

    const networkSshKey = await shallowComponent(<CredentialFormFields typeValue="network" authType="SSH Key" />);
    expect(networkSshKey).toMatchSnapshot('network, SSH Key');
  });

  it('should render specific to authType for type openshift', async () => {
    const openshiftUsernamePassword = await shallowComponent(
      <CredentialFormFields typeValue="openshift" authType="Username and Password" />
    );
    expect(openshiftUsernamePassword).toMatchSnapshot('openshift, "Username and Password"');

    const openshiftToken = await shallowComponent(<CredentialFormFields typeValue="openshift" authType="Token" />);
    expect(openshiftToken).toMatchSnapshot('openshift, "Token"');
  });

  it('should call handlers for setAuthType and handleInputChange', async () => {
    const mockHandleInputChange = jest.fn();
    const mockSetAuthType = jest.fn();
    render(
      <CredentialFormFields
        typeValue="network"
        authType="Username and Password"
        setAuthType={mockSetAuthType}
        handleInputChange={mockHandleInputChange}
      />
    );

    const user = userEvent.setup();

    await user.click(screen.getByText('Username and Password'));
    await user.click(screen.getByText('SSH Key'));
    expect(mockSetAuthType.mock.calls).toMatchSnapshot('setAuthType');

    await user.click(screen.getByText('Select option'));
    await user.click(screen.getByText('sudo'));
    expect(mockHandleInputChange.mock.calls).toMatchSnapshot('handleInputChange');
  });
});

describe('AddCredentialModal-Edit', () => {
  let mockOnClose;
  let mockOnSubmit;
  let mockCredential;

  beforeEach(() => {
    mockOnClose = jest.fn();
    mockOnSubmit = jest.fn();
    mockCredential = {
      id: 1,
      name: 'Test Credential',
      username: 'admin',
      password: '',
      cred_type: 'network',
      created_at: new Date(),
      updated_at: new Date(),
      ssh_key: 'ssh-key',
      auth_token: '',
      ssh_passphrase: '',
      become_method: '',
      become_password: '',
      become_user: '',
      sources: []
    };
    render(
      <AddCredentialModal
        isOpen={true}
        credentialType="network"
        credential={mockCredential}
        onClose={mockOnClose}
        onSubmit={mockOnSubmit}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should have the correct title', () => {
    const title = screen.getByText('Edit Credential');
    expect(title).toMatchSnapshot('AddCredentialModal-edit Title');
  });

  it('should call onSubmit with the correct filtered data when "Save" is clicked', async () => {
    const user = userEvent.setup();
    await user.clear(screen.getByPlaceholderText('Enter a name for the credential'));
    await user.type(screen.getByPlaceholderText('Enter a name for the credential'), 'Test Credential25');
    await user.click(screen.getByText('Save'));

    expect(mockOnSubmit.mock.calls).toMatchSnapshot('onSubmit-edit, filtered data');
  });

  it('should call onClose', async () => {
    const user = userEvent.setup();
    await user.click(screen.getByText('Cancel'));

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });
});
