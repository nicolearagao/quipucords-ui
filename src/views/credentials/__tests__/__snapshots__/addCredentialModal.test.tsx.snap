// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCredentialModal should call onSubmit with the correct filtered data when "Save" is clicked: onSubmit, filtered data 1`] = `
[
  [
    {
      "cred_type": "network",
      "name": "Test Credential",
    },
  ],
]
`;

exports[`AddCredentialModal should have the correct title: AddCredentialModal Title 1`] = `
<span
  class="pf-v5-c-modal-box__title-text"
>
  Add Credential: network
</span>
`;

exports[`AddCredentialModal should render a basic component: basic 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="Add Credential: "
  titleIconVariant={null}
  titleLabel=""
  variant="small"
>
  <Form>
    <CredentialFormFields
      authType="Username and Password"
      formData={
        {
          "auth_token": "",
          "authenticationType": "",
          "become_method": "",
          "become_password": "",
          "become_user": "",
          "name": "",
          "password": "",
          "ssh_key": "",
          "ssh_passphrase": "",
          "username": "",
        }
      }
      handleInputChange={[Function]}
      setAuthType={[Function]}
      typeValue=""
    />
    <ActionGroup>
      <Button
        onClick={[Function]}
        variant="primary"
      >
        Save
      </Button>
      <Button
        onClick={[Function]}
        variant="link"
      >
        Cancel
      </Button>
    </ActionGroup>
  </Form>
</Modal>
`;

exports[`CredentialFormFields should call handlers for setAuthType and handleInputChange: handleInputChange 1`] = `
[
  [
    "become_method",
    "sudo",
  ],
]
`;

exports[`CredentialFormFields should call handlers for setAuthType and handleInputChange: setAuthType 1`] = `
[
  [
    "SSH Key",
  ],
]
`;

exports[`CredentialFormFields should render a basic component: basic 1`] = `
<React.Fragment>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="credential-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter a name for the credential"
      type="text"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_type"
    label="Authentication Type"
  >
    <SimpleDropdown
      dropdownItems={
        [
          "Username and Password",
          "SSH Key",
        ]
      }
      isFullWidth={true}
      label="Username and Password"
      onSelect={[Function]}
      variant="default"
    />
  </FormGroup>
  <React.Fragment>
    <FormGroup
      fieldId="username"
      isRequired={true}
      label="Username"
    >
      <TextInput
        id="credential-username"
        isRequired={true}
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
      />
    </FormGroup>
    <FormGroup
      fieldId="password"
      isRequired={true}
      label="Password"
    >
      <TextInput
        id="credential-password"
        isRequired={true}
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        type="password"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_method"
      label="Become Method"
    >
      <SimpleDropdown
        dropdownItems={
          [
            "Select option",
            "sudo",
            "su",
            "pbrun",
            "pfexec",
            "doas",
            "dzdo",
            "ksu",
            "runas",
          ]
        }
        isFullWidth={true}
        label="Select option"
        onSelect={[Function]}
        variant="default"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_user"
      label="Become User"
    >
      <TextInput
        id="become_user"
        name="become_user"
        onChange={[Function]}
        placeholder="Enter become user (optional)"
        type="text"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_password"
      label="Become Password"
    >
      <TextInput
        id="become_password"
        name="become_password"
        onChange={[Function]}
        placeholder="Enter become password (optional)"
        type="password"
      />
    </FormGroup>
  </React.Fragment>
</React.Fragment>
`;

exports[`CredentialFormFields should render specific to authType for type network: network, SSH Key 1`] = `
<React.Fragment>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="credential-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter a name for the credential"
      type="text"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_type"
    label="Authentication Type"
  >
    <SimpleDropdown
      dropdownItems={
        [
          "Username and Password",
          "SSH Key",
        ]
      }
      isFullWidth={true}
      label="SSH Key"
      onSelect={[Function]}
      variant="default"
    />
  </FormGroup>
  <React.Fragment>
    <FormGroup
      fieldId="username"
      isRequired={true}
      label="Username"
    >
      <TextInput
        id="credential-username"
        isRequired={true}
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
      />
    </FormGroup>
    <FormGroup
      fieldId="ssh_key"
      isRequired={true}
      label="SSH Key"
    >
      <TextArea
        id="credential-ssh-key"
        isRequired={true}
        name="ssh_key"
        onChange={[Function]}
        placeholder="Enter SSH Key"
        rows={10}
      />
    </FormGroup>
    <FormGroup
      fieldId="ssh_passphrase"
      label="SSH passphrase"
    >
      <TextInput
        id="ssh_passphrase"
        name="ssh_passphrase"
        onChange={[Function]}
        placeholder="Enter SSH passphrase (optional)"
        type="password"
      />
    </FormGroup>
  </React.Fragment>
</React.Fragment>
`;

exports[`CredentialFormFields should render specific to authType for type network: network, Username and Password 1`] = `
<React.Fragment>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="credential-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter a name for the credential"
      type="text"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_type"
    label="Authentication Type"
  >
    <SimpleDropdown
      dropdownItems={
        [
          "Username and Password",
          "SSH Key",
        ]
      }
      isFullWidth={true}
      label="Username and Password"
      onSelect={[Function]}
      variant="default"
    />
  </FormGroup>
  <React.Fragment>
    <FormGroup
      fieldId="username"
      isRequired={true}
      label="Username"
    >
      <TextInput
        id="credential-username"
        isRequired={true}
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
      />
    </FormGroup>
    <FormGroup
      fieldId="password"
      isRequired={true}
      label="Password"
    >
      <TextInput
        id="credential-password"
        isRequired={true}
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        type="password"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_method"
      label="Become Method"
    >
      <SimpleDropdown
        dropdownItems={
          [
            "Select option",
            "sudo",
            "su",
            "pbrun",
            "pfexec",
            "doas",
            "dzdo",
            "ksu",
            "runas",
          ]
        }
        isFullWidth={true}
        label="Select option"
        onSelect={[Function]}
        variant="default"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_user"
      label="Become User"
    >
      <TextInput
        id="become_user"
        name="become_user"
        onChange={[Function]}
        placeholder="Enter become user (optional)"
        type="text"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_password"
      label="Become Password"
    >
      <TextInput
        id="become_password"
        name="become_password"
        onChange={[Function]}
        placeholder="Enter become password (optional)"
        type="password"
      />
    </FormGroup>
  </React.Fragment>
</React.Fragment>
`;

exports[`CredentialFormFields should render specific to authType for type openshift: openshift, "Token" 1`] = `
<React.Fragment>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="credential-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter a name for the credential"
      type="text"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_type"
    label="Authentication Type"
  >
    <SimpleDropdown
      dropdownItems={
        [
          "Token",
          "Username and Password",
        ]
      }
      isFullWidth={true}
      label="Token"
      onSelect={[Function]}
      variant="default"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_token"
    isRequired={true}
    label="Token"
  >
    <TextInput
      id="credential-token"
      isRequired={true}
      name="auth_token"
      onChange={[Function]}
      placeholder="Enter Token"
      type="text"
    />
  </FormGroup>
</React.Fragment>
`;

exports[`CredentialFormFields should render specific to authType for type openshift: openshift, "Username and Password" 1`] = `
<React.Fragment>
  <FormGroup
    fieldId="name"
    isRequired={true}
    label="Name"
  >
    <TextInput
      id="credential-name"
      isRequired={true}
      name="name"
      onChange={[Function]}
      placeholder="Enter a name for the credential"
      type="text"
    />
  </FormGroup>
  <FormGroup
    fieldId="auth_type"
    label="Authentication Type"
  >
    <SimpleDropdown
      dropdownItems={
        [
          "Token",
          "Username and Password",
        ]
      }
      isFullWidth={true}
      label="Username and Password"
      onSelect={[Function]}
      variant="default"
    />
  </FormGroup>
  <React.Fragment>
    <FormGroup
      fieldId="username"
      isRequired={true}
      label="Username"
    >
      <TextInput
        id="credential-username"
        isRequired={true}
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
      />
    </FormGroup>
    <FormGroup
      fieldId="password"
      isRequired={true}
      label="Password"
    >
      <TextInput
        id="credential-password"
        isRequired={true}
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        type="password"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_method"
      label="Become Method"
    >
      <SimpleDropdown
        dropdownItems={
          [
            "Select option",
            "sudo",
            "su",
            "pbrun",
            "pfexec",
            "doas",
            "dzdo",
            "ksu",
            "runas",
          ]
        }
        isFullWidth={true}
        label="Select option"
        onSelect={[Function]}
        variant="default"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_user"
      label="Become User"
    >
      <TextInput
        id="become_user"
        name="become_user"
        onChange={[Function]}
        placeholder="Enter become user (optional)"
        type="text"
      />
    </FormGroup>
    <FormGroup
      fieldId="become_password"
      label="Become Password"
    >
      <TextInput
        id="become_password"
        name="become_password"
        onChange={[Function]}
        placeholder="Enter become password (optional)"
        type="password"
      />
    </FormGroup>
  </React.Fragment>
</React.Fragment>
`;

exports[`useCredentialForm should derive token auth for specific credential types: token auth 1`] = `
{
  "openshift": "Token",
  "rhacs": "Token",
}
`;

exports[`useCredentialForm should initialize formData correctly: formData 1`] = `
{
  "auth_token": "",
  "authenticationType": "",
  "become_method": "",
  "become_password": "",
  "become_user": "",
  "name": "",
  "password": "",
  "ssh_key": "",
  "ssh_passphrase": "",
  "username": "",
}
`;
